rust/core-lib/src/file.rs

```diff
@@ -172,6 +190,79 @@ impl FileManager {
         }
         Ok(())
     }
+
+    #[cfg(feature = "notify")]
+    pub fn update_current_position_in_tail(
+        &mut self,
+        buffer_id: BufferId,
+    ) -> Result<bool, FileError> {
+        if let Some(file_info) = self.file_info.get_mut(&buffer_id) {
+            let path = &file_info.path.to_owned();
+
+            let mut f = File::open(path).map_err(|e| FileError::Io(e, path.to_owned()))?;
+            let end_position =
+                f.seek(SeekFrom::End(0)).map_err(|e| FileError::Io(e, path.to_owned()))?;
+
+            file_info.tail_position = Some(end_position);
+            self.watcher.unwatch(path, OPEN_FILE_EVENT_TOKEN);
+            self.watcher.watch(path, false, OPEN_FILE_TAIL_EVENT_TOKEN);
+            Ok(true)
+        } else {
+            info!("Non-existent file cannot be tailed!");
+            Ok(false)
+        }
+    }
+
+    #[cfg(feature = "notify")]
+    pub fn disable_tailing(&mut self, buffer_id: BufferId) -> bool {
+        if let Some(file_info) = self.file_info.get_mut(&buffer_id) {
+            let path = &file_info.path.to_owned();
+            file_info.tail_position = None;
+            self.watcher.unwatch(path, OPEN_FILE_TAIL_EVENT_TOKEN);
+            self.watcher.watch(path, false, OPEN_FILE_EVENT_TOKEN);
+            true
+        } else {
+            info!("Non-existent file cannot be tailed!");
+            false
+        }
+    }
+}
+
+/// When tailing a file, instead of reading file from beginning, we need to get changes from the end.
+/// This method does that.
+#[cfg(feature = "notify")]
+fn try_tailing_file<P>(path: P, file_info: &FileInfo) -> Result<(Rope, FileInfo), FileError>
+where
+    P: AsRef<Path>,
+{
+    let mut f =
+        File::open(path.as_ref()).map_err(|e| FileError::Io(e, path.as_ref().to_owned()))?;
+    let end_position =
+        f.seek(SeekFrom::End(0)).map_err(|e| FileError::Io(e, path.as_ref().to_owned()))?;
+
+    let current_position = file_info.tail_position.expect(
+        "tail_position is None. Since we are tailing a file, tail_position cannot be None.",
+    );
+    let diff = end_position - current_position;
+    let mut buf = vec![0; diff as usize];
+    f.seek(SeekFrom::Current(-(buf.len() as i64))).unwrap();
```

Comment:
Isn't buf.len() == diff here?

Reply:
Also, it'd be nice if we could avoid unwrap

Reply:
> Isn't buf.len() == diff here?

yes.

Reply:
So it'd make more sense to just use diff here directly I think

Reply:
Ok. Will make the change.